### ISSUE TITLE
anyhow's dyn error `backtrace` disagrees with `anyhow::Error::backtrace`

### ISSUE THREAD

**Issue by guswynn** [dtolnay/anyhow#231]
```
#![feature(backtrace)]
use anyhow;
use thiserror::Error;

#[derive(Error, Debug)]
enum Gus {
    #[error("hmm")]
    Other(
        #[from]
        #[backtrace]
        anyhow::Error,
    ),
}
fn main() {
    use std::error::Error;
    use std::ops::Deref;
    println!("{:?}", Gus::from(anyhow::anyhow!("main")));
    println!("{:?}", Gus::from(anyhow::anyhow!("main")).backtrace());
    println!("{:?}", anyhow::anyhow!("main").deref().backtrace());
}
  ```
  
  results in 
```
>printing the error with a backtrace
None
None
```

Looking briefly at the code, this appears to be at least semi-on-purpose, where the vtable for adhoc errors points to `no_backtrace`, but it links the backtrace back in somehow? can the `ErrorImpl` coercion do the same things as https://github.com/dtolnay/anyhow/blob/6833150a7965ad9f489a1da619298ba84b8d3148/src/error.rs#L880?

(also, is there a reason the backtrace is only printed for the `Debug` impl, not the `Display` impl, should the docs be updated to reflect that?)

**Comment by dtolnay** [comment]
Updated repro:

```rust
#![feature(error_generic_member_access, provide_any)]

use std::any;
use std::backtrace::Backtrace;
use std::ops::Deref;
use thiserror::Error;

#[derive(Error, Debug)]
enum Gus {
    #[error("hmm")]
    Other(
        #[from]
        #[backtrace]
        anyhow::Error,
    ),
}

fn main() {
    println!("{:?}", Gus::from(anyhow::anyhow!("main")));
    println!("{:?}", any::request_ref::<Backtrace>(&Gus::from(anyhow::anyhow!("main"))));
    println!("{:?}", any::request_ref::<Backtrace>(&anyhow::anyhow!("main").deref()));
}
```

```console
Other(main

Stack backtrace:
   …)
None
None
```

**Comment by dtolnay** [comment]
This is fixed as of thiserror 1.0.34 and anyhow 1.0.64.

```console
Other(main

Stack backtrace:
   …)
Some(Backtrace […])
None
```

The last one is still `None` because you've written `.deref()` which specifically dereferences to the underlying Error impl that the anyhow::Error was constructed from, but the thiserror-generated Error impl has the correct backtrace now. It may be possible to do it for `.deref()` too but since your issue involves thiserror it doesn't sound like your issue was motivated by that case.
